x/oracle/keeper/historic_price.go

```diff
@@ -0,0 +1,201 @@
+func (k Keeper) setMedian(
+	ctx sdk.Context,
+	denom string,
+) {
+	store := ctx.KVStore(k.storeKey)
+
+	historicPrices := k.GetHistoricPrices(ctx, denom)
+
+	median := median(historicPrices)
+
+	bz := k.cdc.MustMarshal(&sdk.DecProto{Dec: median})
+	denom = strings.ToUpper(denom)
+	store.Set(types.GetMedianKey(denom), bz)
+}
+
+// GetHistoricPrice returns the historic price of a denom at a given
+// block.
+func (k Keeper) GetHistoricPrice(
+	ctx sdk.Context,
+	denom string,
+	blockNum uint64,
+) (types.HistoricPrice, error) {
+	store := ctx.KVStore(k.storeKey)
+	denom = strings.ToUpper(denom)
+	bz := store.Get(types.GetHistoricPriceKey(denom, blockNum))
+	if bz == nil {
+		return types.HistoricPrice{}, sdkerrors.Wrap(types.ErrUnknownDenom, denom)
+	}
+
+	var historicPrice types.HistoricPrice
+	k.cdc.MustUnmarshal(bz, &historicPrice)
+
+	return historicPrice, nil
+}
+
+// GetHistoricPrices returns all the historic prices of a given denom.
+func (k Keeper) GetHistoricPrices(
+	ctx sdk.Context,
+	denom string,
+) []types.HistoricPrice {
+	denom = strings.ToUpper(denom)
+	historicPrices := []types.HistoricPrice{}
+
+	k.IterateHistoricPrices(ctx, denom, func(exchangeRate sdk.Dec, blockNum uint64) bool {
+		historicPrices = append(historicPrices, types.HistoricPrice{
+			ExchangeRates: types.ExchangeRateTuple{
+				Denom:        denom,
+				ExchangeRate: exchangeRate,
+			},
+			BlockNum: blockNum,
+		})
+
+		return false
+	})
+
+	return historicPrices
+}
+
+// IterateHistoricPrices iterates over historic prices of a given
+// denom in the store.
+func (k Keeper) IterateHistoricPrices(
+	ctx sdk.Context,
+	denom string,
+	handler func(sdk.Dec, uint64) bool,
+) {
+	store := ctx.KVStore(k.storeKey)
+
+	iter := sdk.KVStorePrefixIterator(store, types.KeyPrefixHistoricPrice)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		key := iter.Key()
+		historicPriceDenom := string(key[len(types.KeyPrefixExchangeRate) : len(key)-9])
+		if historicPriceDenom != denom {
+			continue
+		}
+		historicPrice := types.HistoricPrice{}
+
+		k.cdc.MustUnmarshal(iter.Value(), &historicPrice)
+		if handler(historicPrice.ExchangeRates.ExchangeRate, historicPrice.BlockNum) {
+			break
+		}
+	}
+}
+
+// AddHistoricPrice adds the historic price of a denom at the current
+// block height when called to the store. Afterwards it will call
+// setMedian to update the median price of the denom in the store.
+func (k Keeper) AddHistoricPrice(
+	ctx sdk.Context,
+	denom string,
+	exchangeRate sdk.Dec,
+) {
+	store := ctx.KVStore(k.storeKey)
+	exchangeRateTuple := types.ExchangeRateTuple{
+		Denom:        denom,
+		ExchangeRate: exchangeRate,
+	}
+
+	block := uint64(ctx.BlockHeight())
+	bz := k.cdc.MustMarshal(&types.HistoricPrice{
+		ExchangeRates: exchangeRateTuple,
+		BlockNum:      block,
+	})
+	denom = strings.ToUpper(denom)
+	store.Set(types.GetHistoricPriceKey(denom, block), bz)
+
+	// update median after adding new historic price
+	k.setMedian(ctx, denom)
```

Comment:
You'll want to pass in the exchangeRate rather than querying for it again in `setMedian`
```suggestion
	k.setMedian(ctx, denom, exchangeRate)
```